storage/storage_image.go

```
@@ -90,309 +95,654 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
+	i := s.layerPosition[info.Digest]
+	s.layerPosition[info.Digest] = i + 1
+	if len(uncompressedLayers) > 0 {
+		layer = uncompressedLayers[i%len(uncompressedLayers)]
+	} else {
+		layer = compressedLayers[i%len(compressedLayers)]
+	}
+	layerID = layer.ID
+	// If we've been told to force decompression, do so.
+	if decompress {
+		noCompression := archive.Uncompressed
+		diffOptions = &storage.DiffOptions{
+			Compression: &noCompression,
+		}
+		if layer.UncompressedSize < 0 {
+			n = -1
+		} else {
+			n = layer.UncompressedSize
+		}
+		logrus.Debugf("exporting filesystem layer %q without compression for blob %q", layerID, info.Digest)
+	} else {
+		if layer.CompressedSize < 0 {
+			n = -1
+		} else {
+			n = layer.CompressedSize
+		}
+		logrus.Debugf("exporting filesystem layer %q with default compression (%v) for blob %q", layer.CompressionType, layerID, info.Digest)
+	}
+	rc, err = s.transport.store.Diff("", layerID, diffOptions)
+	if err != nil {
+		return nil, -1, "", err
+	}
+	return rc, n, layerID, err
+}
+
+// GetManifest() reads the image's manifest.
+func (s *storageUnnamedImageReader) GetManifest() (manifestBlob []byte, MIMEType string, err error) {
+	manifestBlob, err = s.transport.store.ImageBigData(s.ID, "manifest")
+	return manifestBlob, manifest.GuessMIMEType(manifestBlob), err
+}
+
+// GetTargetManifest() is not supported.
+func (s *storageUnnamedImageReader) GetTargetManifest(d digest.Digest) (manifestBlob []byte, MIMEType string, err error) {
+	return nil, "", ErrNoManifestLists
+}
+
+// GetSignatures() parses the image's signatures blob into a slice of byte slices.
+func (s *storageUnnamedImageReader) GetSignatures() (signatures [][]byte, err error) {
+	var offset int
+	sigslice := [][]byte{}
+	signature, err := s.transport.store.ImageBigData(s.ID, "signatures")
+	if err != nil {
+		if !os.IsNotExist(errors.Cause(err)) {
+			logrus.Debugf("got error %v looking up signatures for image %q", err, s.ID)
```

A file-not-found error would be ignored, but I'll restructure it to not do that, and instead skip the attempt to read a file if we don't expect to find anything.